R/particle_filter_state.R

```diff
@@ -31,28 +31,134 @@ particle_filter_state <- R6::R6Class(
+      history_order <- history$order
 
-      model <- self$model
+      log_likelihood <- self$log_likelihood
+      n_particles <- private$n_particles
+
+      min_log_likelihood <- private$min_log_likelihood
+      support <- private$support
+
+      for (t in seq(curr + 1L, step_index)) {
+        step_end <- steps[t, 2L]
+        state <- model$run(step_end)
+
+        if (save_history) {
+          ## NOTE: There are two places here (and for the order below)
+          ## where we assign trajectories, and we have to do this
+          ## without using `array_last_dimension<-`, otherwise there's
+          ## a big performance regression due to excessive GC (we make
+          ## a copy into the function call, I suspect?)
+          ##
+          ## An alternative approach here would be to store the
+          ## history as a flat array (or access it as such), then
+          ## compute what the index would be.  If we know the product
+          ## of the first dimensions, this is pretty easy really.
+          if (nested) {
+            history_value[, , , t + 1L] <- model$state(save_history_index)
+          } else {
+            history_value[, , t + 1L] <- model$state(save_history_index)
+          }
+        }
+
+        weights <- support$compare(state, compare, data_split[[t]], pars)
+
+        if (is.null(weights)) {
+          log_likelihood_step <- NA_real_
+          kappa <- seq_len(n_particles)
+        } else {
+          log_likelihood_step <- weights$average
+          log_likelihood <- log_likelihood + log_likelihood_step
+
+          if (particle_filter_early_exit(log_likelihood, min_log_likelihood)) {
+            log_likelihood[] <- -Inf
+            break
+          }
+
+          kappa <- particle_resample(weights$weights)
+          model$reorder(kappa)
+        }
+
+        if (save_history) {
+          if (nested) {
+            history_order[, , t + 1L] <- kappa
+          } else {
+            history_order[, t + 1L] <- kappa
+          }
+        }
+
+        if (save_restart) {
+          i_restart <- match(step_end, save_restart_step)
+          if (!is.na(i_restart)) {
+            array_last_dimension(restart_state, i_restart) <- model$state()
+          }
+        }
+      }
+
+      self$log_likelihood_step <- log_likelihood_step
+      self$log_likelihood <- log_likelihood
+      self$current_step_index <- step_index
       if (save_history) {
-        model$set_index(save_history_index)
-        on.exit(model$set_index(integer(0)))
+        history$value <- history_value
+        history$order <- history_order
+        self$history <- history
+      }
+      if (save_restart) {
+        self$restart_state <- restart_state
+      }
+
+      if (partial) {
+        log_likelihood_step
+      } else {
+        log_likelihood
       }
+    },
+
+    step_compiled = function(step_index, partial = FALSE) {
+      if (partial) {
+        stop("'partial' not supported with compiled compare")
```

Comment:
Maybe I'm confused, but I thought support was just added for this? Or am I misunderstanding what `partial` is?

Reply:
partial is the support we put in for (one of) if2/smc2, where you get the likelihood contributed by the last single step. We don't support that but we do support running to a point that is not the end. It's confusing yes :)

Reply:
partial is the support we put in for (one of) if2/smc2, where you get the likelihood contributed by the last single step. We don't support that but we do support running to a point that is not the end. It's confusing yes :)